### ISSUE TITLE
[Serializable] attribute is not produced in metadata generation

### ISSUE THREAD

**Issue by AArnott** [dotnet/roslyn#3898]
When I hit F12 on a reference to the `System.Reflection.AssemblyName` type, the top of the generated file that is opened in the editor is:

``` csharp
#region Assembly mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
// C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\mscorlib.dll
#endregion

using System.Configuration.Assemblies;
using System.Globalization;
using System.Runtime.InteropServices;
using System.Runtime.Serialization;
using System.Security;

namespace System.Reflection
{
    [ClassInterface(ClassInterfaceType.None)]
    [ComDefaultInterface(typeof(_AssemblyName))]
    [ComVisible(true)]
    public sealed class AssemblyName : _AssemblyName, ICloneable, ISerializable, IDeserializationCallback
    {
```

Note the lack of the `[Serializable]` attribute on the type. But the attribute belongs there as that is what was in the original source file. It seems this attribute is special (it compiles to a different kind of metadata on the type), and requires special care when decompiling in order to reproduce it.

**Comment by jaredpar** [comment]
@AArnott I think this is actually behaving correctly.  Or at least the reader is functioning correctly for that specific DLL.  If you open up the referenced assembly at the top of the file in ILDASM you'll see that `AssemblyName` isn't annotated with `[Serializable]`.  It looks like it was excluded from that particular reference assembly. 

It is still possible though that the IDE is picking the wrong assembly to use a metadata view on though.

**Comment by tmat** [comment]
I see this in ILDASM:
```.class public auto ansi ***serializable*** sealed beforefieldinit System.Reflection.AssemblyName```

**Comment by jaredpar** [comment]
@tmat ah. Didn't realize it was hiding out as a flag vs. an attribute.

**Comment by tmat** [comment]
Yeah, it seems that we're not exposing (some) pseudo-custom attributes thru symbol APIs as regular attributes. I remember there was a discussion on this topic a while ago and we decided to expose them but not sure if it ever got implemented.

**Comment by tmat** [comment]
More precisely - those pseudo-custom attributes that the language doesn't have a dedicated syntax for should be presented as custom attributes.

**Comment by Pilchie** [comment]
Moving to compiler, based on @tomat's comments above.

**Comment by tomat** [comment]
You probably meant @tmat :-)

**Comment by Pilchie** [comment]
Oops, you're right of course.

**Comment by gafter** [comment]
@Pilchie While I agree with @tmat's assessment, the code the user is seeing isn't produced by the compiler, is it? If it is, can you please identify the offending component and reassign to the compiler team?

**Comment by gafter** [comment]
Right, the attribute should appear in the symbol API.

**Comment by lgolding** [comment]
Hi @AArnott (and @bkoelman, who has also reported this issue),

The compiler team will address this problem by "synthesizing" a `[Serializable]` attribute for any type that has the "serializable" bit set in the metadata.

We hope to address this in VS 2015 Update 3.

Hope this helps,
Larry

@srivatsn @jaredpar @mattwar

**Comment by lgolding** [comment]
(This has been reported a few times. We will use this issue, roslyn/#3898, to track it.)

**Comment by lgolding** [comment]
Once this is fixed, it will no longer be necessary for the CA2235 analyzer to special-case the primitive types `int`, `char`, _etc._, because the underlying CLR types such as `System.Int32` are all defined to be serializable. +@nguerrera.

**Comment by bkoelman** [comment]
Nice, thanks!

**Comment by JasonBock** [comment]
Is this going to be in 15.6 or 15.7? This has been open for over two years now. Any chance of this being resolved?

**Comment by jaredpar** [comment]
@JasonBock this hasn't been done yet because it just hasn't been high priority enough. Is there a particular scenario driving your ask?

**Comment by JasonBock** [comment]
@jaredpar yes, [this issue](https://github.com/MarimerLLC/csla/issues/623). I can't write an analyzer that can check the arguments to a method to see if they're serializable or not, at least I can't figure out a way to do it without this issue being addressed.

**Comment by jcouv** [comment]
@jaredpar I moved this issue back to 15.7, as it was also hit by https://github.com/dotnet/roslyn-analyzers/issues/1510 (reported to us in https://github.com/dotnet/roslyn/issues/24133). Let me know if that's ok.

**Comment by jcouv** [comment]
Started a PR to fix this. It still has some limitations that I'll try to fix.

**Comment by jcouv** [comment]
Fix is merged and should ship in 15.7. It adds an `INamedTypeSymbol.IsSerializable` API.
